protocol-units/bridge/chains/ethereum/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+use bridge_shared::types::{
+	Amount, BridgeTransferDetails, BridgeTransferId, HashLock, HashLockPreImage, InitiatorAddress,
+	RecipientAddress, TimeLock,
+};
+use bridge_shared::{
+	bridge_contracts::{
+		BridgeContractCounterpartyError, BridgeContractInitiator, BridgeContractInitiatorError,
+		BridgeContractInitiatorResult,
+	},
+	bridge_monitoring::{BridgeContractCounterpartyEvent, BridgeContractInitiatorEvent},
+};
+use keccak_hash::keccak;
+use mcr_settlement_client::send_eth_transaction::{
+	InsufficentFunds, SendTransactionErrorRule, UnderPriced, VerifyRule,
+};
+
+use std::fmt::Debug;
+use utils::{send_transaction, EthAddress};
+
+pub mod utils;
+
+const INITIATOR_ADDRESS: &str = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
+const RECIPIENT_ADDRESS: &str = "0xf39Fd6e51aad88F6F4ce6aB8827279cffFb92266";
+const DEFAULT_GAS_LIMIT: u64 = 10_000_000_000;
+const MAX_RETRIES: u32 = 5;
+
+type EthHash = [u8; 32];
+
+pub type SCIResult<A, H> = Result<BridgeContractInitiatorEvent<A, H>, BridgeContractInitiatorError>;
+pub type SCCResult<A, H> =
+	Result<BridgeContractCounterpartyEvent<A, H>, BridgeContractCounterpartyError>;
+
+///Configuration for the Ethereum Bridge Client
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Config {
+	pub rpc_url: Option<String>,
+	pub ws_url: Option<String>,
+	pub chain_id: String,
+	pub signer_private_key: String,
+	pub initiator_address: EthAddress,
+	pub recipient_address: String,
+	pub gas_limit: u64,
+	pub num_tx_send_retries: u32,
+}
+
+impl Default for Config {
+	fn default() -> Self {
+		Config {
+			rpc_url: Some("http://localhost:8545".to_string()),
+			ws_url: Some("ws://localhost:8545".to_string()),
+			chain_id: "31337".to_string(),
+			signer_private_key: Self::default_for_private_key(),
+			initiator_address: EthAddress::from(INITIATOR_ADDRESS.to_string()),
+			recipient_address: RECIPIENT_ADDRESS.to_string(),
+			gas_limit: DEFAULT_GAS_LIMIT,
+			num_tx_send_retries: MAX_RETRIES,
+		}
+	}
+}
+
+impl Config {
+	fn default_for_private_key() -> String {
+		let random_wallet = PrivateKeySigner::random();
+		random_wallet.to_bytes().to_string()
+	}
+}
+
+// Codegen from the abi
+sol!(
+	#[allow(missing_docs)]
+	#[sol(rpc)]
+	AtomicBridgeInitiator,
+	"abis/AtomicBridgeInitiator.json"
+);
+
+type AlloyProvider = FillProvider<
+	JoinFill<
+		JoinFill<
+			JoinFill<JoinFill<alloy::providers::Identity, GasFiller>, NonceFiller>,
+			ChainIdFiller,
+		>,
+		WalletFiller<EthereumWallet>,
+	>,
+	RootProvider<BoxTransport>,
+	BoxTransport,
+	Ethereum,
+>;
+
+#[derive(RlpDecodable, RlpEncodable)]
+struct EthBridgeTransferDetails {
+	pub amount: U256,
+	pub originator: EthAddress,
+	pub recipient: [u8; 32],
+	pub hash_lock: [u8; 32],
+	pub time_lock: U256,
+	pub state: u8, // Assuming the enum is u8 for now..
+}
+
+pub struct EthClient<P> {
+	rpc_provider: P,
+	chain_id: String,
+	ws_provider: RootProvider<PubSubFrontend>,
+	initiator_address: EthAddress,
+	signing_address: EthAddress,
```

Comment:
We need the signer_address to initialise the contract `let contract = AtomicBridgeInitiator::new(self.signing_address.0, &self.rpc_provider);`, this is not always the `initiator_address` because it can be called by any client.